src/v/cloud_storage/topic_recovery_service.cc

```diff
@@ -0,0 +1,599 @@
+    _downloaded_manifests.emplace(manifests);
+
+    populate_recovery_status();
+
+    co_await propagate_state(state::creating_topics);
+    vlog(cst_log.debug, "creating topics");
+
+    result<void, recovery_error_ctx> result = outcome::success();
+    try {
+        auto create_topics_results = co_await create_topics(request);
+        for (const auto& r : create_topics_results) {
+            if (r.ec != cluster::errc::success) {
+                vlog(cst_log.warn, "topic creation failed: {}", r);
+                _recovery_status.erase(r.tp_ns);
+            } else {
+                vlog(
+                  cst_log.debug,
+                  "topic created: {}, expected downloads: {}",
+                  r,
+                  _recovery_status[r.tp_ns].pending_downloads);
+            }
+        }
+
+        co_await propagate_state(state::recovering_data);
+        start_download_bg_tracker();
+    } catch (const ss::timed_out_error& err) {
+        result = recovery_error_ctx::make(
+          fmt::format("failed to create topics: {}", err),
+          recovery_error_code::error_creating_topics);
+    }
+
+    if (result.has_error()) {
+        co_await propagate_state(state::inactive);
+    }
+    co_return result;
+}
+
+static cluster::topic_configuration make_topic_config(
+  const cloud_storage::topic_manifest& tm, const recovery_request& request) {
+    auto topic_config = tm.get_topic_config();
+    if (!topic_config.has_value()) {
+        vlog(
+          cst_log.warn,
+          "skipping topic creation for {}, missing values in manifest",
+          tm.get_manifest_path());
+    }
+
+    cluster::topic_configuration topic_to_create_cfg(
+      topic_config->tp_ns.ns,
+      topic_config->tp_ns.tp,
+      topic_config->partition_count,
+      topic_config->replication_factor);
+
+    auto& topic_properties = topic_to_create_cfg.properties;
+    auto manifest_props = topic_config->properties;
+
+    topic_properties.compression = manifest_props.compression;
+    topic_properties.cleanup_policy_bitflags
+      = manifest_props.cleanup_policy_bitflags;
+    topic_properties.compaction_strategy = manifest_props.compaction_strategy;
+
+    topic_properties.retention_duration = tristate<std::chrono::milliseconds>{
+      default_retention_ms_on_topic_create};
+    topic_properties.retention_local_target_ms
+      = tristate<std::chrono::milliseconds>{
+        default_retention_ms_on_topic_create};
+
+    topic_properties.segment_size = manifest_props.segment_size;
+    topic_properties.timestamp_type = manifest_props.timestamp_type;
+    topic_properties.shadow_indexing = model::shadow_indexing_mode::full;
+    topic_properties.recovery = true;
+
+    if (request.retention_bytes().has_value()) {
+        topic_properties.retention_local_target_bytes = tristate<size_t>{
+          request.retention_bytes()};
+        topic_properties.retention_bytes = tristate<size_t>{
+          request.retention_bytes()};
+    }
```

Comment:
Should we clear `retention_local_target_ms` if the user specified a custom retention size?

Reply:
changed to clear size/duration if the other property is specified.